blog: how reliable is your distributed database (#1435)
questdb/questdb.io: blog/2023-03-03-how-reliable-is-your-distributed-database.mdx

author_image_url: https://avatars.githubusercontent.com/ideoma
description: Comparion of replication options of time-series databases.
keywords:
  - replication
  - failover
  - Write Ahead Log
  - comparison
image: /img/blog/2023-03-03/database-replication.png
tags: [replication, comparison, engineering]
slug: how-reliable-your-distributed-database
---

A brief analysis of what can go wrong with databases and how different
time-series databases can use replication to solve common write failover
scenarios.

<!--truncate-->

## Failover

Last year, I started to commute to the office since the pandemic loosened its
grip on London life. To beat the traffic and unreliable (as well as expensive!)
trains, I bought a quality electric bicycle from a famous Swiss-American brand.
It's a well-thought-out device; sometimes I even want to call it a gadget!
Despite how enjoyable it is, in the last 6 months, I've had to go to the bike
repair shop to fix quite a few things. There were some gearbox glitches (yes, it
has a gearbox), then broken electrical wiring to the backlight stopped the bike
from running. Soon after that, I had to do some trivial maintenance, replacing

> the cut tire and brake pads. Finally, the driving belt snapped (it has a belt

Author: accepted

database also has to evolve the schema to support adding/removing columns from
multiple nodes and applying the transaction in exactly the same order on all the
replicas. This is because some transactions like `INSERT` and `UPDATE` that are
executed on the same row will have different outcomes if applied in the wrong
order.

To summarize, a database needs to be able to:

- Write data to the same table using multiple DB Nodes
  ([multi-master replication](https://en.wikipedia.org/wiki/Multi-master_replication))
  or mirror the data to a read-only replicated node with automatic failover

- Evolve table schemas such that the same table columns can be added
  simultaneously from multiple connections

- Maintain the global order of the writes and schema changes across replica
  nodes

## Sync and Async replication

Different things can go wrong with databases when a failure happens. For
instance, there can be two reasons why the database `Node 1` may not reply to
`insert into A values(2)`: it can either be that the node fails to receive and
process the transaction, or that the OK reply is not delivered back to the
application due to a network disconnect.

> To avoid losing the second insert after disconnecting from `Node 1`, the

Author: rewrote the paragraph in very detailed way

application has to send `insert into A values(2)` a second time to `Node 2` so
that the database does not create a duplicate record. In other words, data
inserted into `Node 1` has to be readable from `Node 2`  
immediately for the application to perform the deduplication. Alternatively,
there must be another built-in mechanism to de-duplicate the inserted rows in
the database. This leads us to the conclusion that at least one of the points
below has to exist for a non-duplicating write failover:

- Data written to one node is immediately selectable through all other replicas
  so that the writing application can check for the duplicates

- There is a de-duplication mechanism built into the insert protocol or table
  storage

The first bullet point above is also called Synchronous (Sync) Replication where
`Node 1` replies `OK` only after an insert is already replicated to `Node 2`. In
contrast, Asynchronous (Async) replication allows `Node 1` to reply `OK` to the
`App` before replicating the inserted data to `Node 2`. This way `Node 2` may
not receive the first insert before `Node 1` goes down, so the database will

> have to reattempt to replicate the data when `Node 1` is started/connected back.

Author: accpted

simultaneously trying to roll another alongside you at all times. You can
imagine that it's pretty hard to ride while simultaneously balancing on two
bikes, and that it's nearly impossible to actually cycle quickly!

<Banner
  alt="Sync bike riding"
  src="/img/blog/2023-03-03/bikes_alongside.jpeg"
  width={650}
>
  Sync replication bike riding
</Banner>

On the other hand, a database configured in single primary with a read-only
Async replica is similar to me buying a spare bike, storing it at work, and
synchronizing my trip data when the bikes are at the same place. In case of a
breakdown, I'll lose data from that particular journey but I would still be able
to continue commuting on the spare bike the same day.

I imagine that multi-master replication is something like riding with a friend
on two tandem bikes, where everyone is riding the tandem on the first seat. If
one of the tandems breaks, the rider can move to the back seat of the other one.
Multi-master Sync replication is riding two tandems next to each other without
the freedom of turning or stopping independently. Async replication would be the
independent rides with occasional location/trip data catch-ups.

> Async replication is the most reasonable choice in the world of bicycles, I do

Author: accepted

```SQL
INSERT INTO A VALUES(1);
```

And `Alice` sends to `Node 2`

```SQL
INSERT INTO A VALUES(2);
```

Both records will be written to each of the nodes. When the
`INSERT INTO A VALUES(1)` statement is received on `Node 1`, ClickHouse writes
it to `part 1_1`. Next, `Node 1` registers the data part with the Zookeeper (or
ClickHouse Keeper). Zookeeper notifies each node about the new part, and the
nodes download the data from the source and apply it to the local table replica.
The same process happens simultaneously with `INSERT 2`.

<Banner
  alt="ClickHouse multi-master replication"
  src="/img/blog/2023-03-03/Multi-maser-replication-ClickHouse.png"
  width={744}
>
  ClickHouse multi-master replication
</Banner>

In this architecture, inserts can be written to each of the nodes in parallel.

What about reading the data back? ClickHouse documentation states that the
replication process is Asynchronous and it may take some time for `Node 2` to
catch up with `Node 1`. There is however an option to specify `insert_quorum`
with every insert. If the `insert_quorum` is set to `2` then the application
gets confirmation back from the database after both `Node 1` and `Node 2` have

> the data inserted, effectively turning this into Sync replication. There are a

Reviewer: ```suggestion
inserted the data, effectively turning this into Sync replication. There are a
```

Author: accepted

  ([multi-master replication](https://en.wikipedia.org/wiki/Multi-master_replication))
  or mirror the data to a read-only replicated node with automatic failover

- Evolve table schemas such that the same table columns can be added
  simultaneously from multiple connections

- Maintain the global order of the writes and schema changes across replica
  nodes

## Sync and Async replication

Different things can go wrong with databases when a failure happens. For
instance, there can be two reasons why the database `Node 1` may not reply to
the second insert, `insert into A values(2)`: it can either be that the node fails to receive and
process the transaction, or that the data is inserted but the OK reply is not delivered back to the
application due to a network disconnect.

To avoid losing the second insert after disconnecting from `Node 1`, the
application has to repeat transaction `insert into A values(2)` to `Node 2`.
The application also has to do the repeated `insert into A values(2)` attempt
in such a way that the database does not create a duplicate row
in the case when the same insert has already been processed by `Node 1`

> but OK reply is not delivered back to the application.

Reviewer: ```suggestion
but an OK reply is not delivered back to the application.
```

> In order to achieve this, all data inserted into `Node 1` has to be readable from `Node 2`

Reviewer 2: Here we talk about two things so:

```suggestion
In order to achieve these goals, all data inserted into `Node 1` has to be readable from `Node 2`
```

[and so does Timescale](https://docs.timescale.com/timescaledb/latest/how-to-guides/replication-and-ha/).

Unfortunately, automatic failover is solved neither by PostgreSQL nor
TimescaleDB, but there are 3rd-party solutions like
[Patroni](https://github.com/zalando/patroni) that add support for that
functionality. PostgreSQL describes the process
[of failover as STONITH (Shoot The Other Node In The Head)](https://www.postgresql.org/docs/15/warm-standby-failover.html),
meaning that the primary node has to be shot down once it starts to misbehave.

Running Sync replication can solve the data gaps and duplicate problems after
the failover. If the application detects the failover, it can re-run the last
non-confirmed `INSERT` as an [UPSERT](https://wiki.postgresql.org/wiki/UPSERT).
With Async replication, a few recent transactions may be missing on the replica
that is promoted to primary. This is because the old primary node had to be shot
down (STONISHed) and there is no trivial way to move the missing data from that
"dead" node to the new primary node post-failover.

## ClickHouse

ClickHouse had been developed Open Source for many years by Yandex, a search

> provider in Russia. ClickHouse's functionality in open source (Apache 2.0)

Reviewer: "Open Source" vs "open source" capitalization should be consistent

Author: lower case every where now

the most popular approach being `ReplicatedMergeTree` storage.

## InfluxDB

InfluxDB has the highest
[DB engines time series ranking](https://db-engines.com/en/ranking/time+series+dbms)
at the time of writing, and I feel that it has to be included here even though
InfluxData
[removed the clustering product from the open source version in 2016](https://en.wikipedia.org/wiki/InfluxDB#Closed_source_clustering_components)
to sell it as a commercial product, InfluxDB Enterprise. Since then, their focus
has shifted from the enterprise version to the cloud offering in recent years,
where they have built InfluxDB Cloud v2. While this is a closed-source system,
its high-level architecture is deducible from the marketing diagrams InfluxData
officially provides.

<Banner
  alt="InfluxDB cloud architecture"
  src="/img/blog/2023-03-03/Influx Cloud Architecture.png"
  width={750}
>
  InfluxDB cloud architecture
</Banner>

InfluxDB Cloud v2 persists incoming writes to the Write Ahead Log (WAL) written
over a Kafka cluster. It is a clean solution that solves the durability and
distribution of the WAL and ensures that the data is already replicated when the
client receives a write confirmation.

> Applying WAL to the "Queryable" storage runs asynchronously in the Ingester
> component, consuming messages from Kafka WAL and writing them to 2 independent
> TSDB copies. There is a delay between reading the message confirmed to be

Reviewer 2: I am guessing that WAL is consuming the Kafka messages:

```suggestion
WAL is applied to the "Queryable" storage, running asynchronously in the Ingester
component, consuming messages from the Kafka WAL, and writing them to 2 independent
TSDB copies. There is a delay between reading the message confirmed to be
```

Grammatically speaking, "Applying WAL to the "Queryable" storage" is the subject but this subject, which is an action, cannot really consume Kafka messages and writes the messages, hence my rewriting.

Author: Looking at it, I just want to remove WAL word from `Kafka WAL` and the rest looks better as it is.

Reviewer 2: For me this works:

WAL application to "Queriable" table storage runs asynchronously in the Ingester component, consuming messages from Kafka and writing them to 2 independent
TSDB copies.

The subject "applying WAL..." does not work in the sentence structure.

Author: accepted the latest, thanks

> written, in Influx term this is called

Reviewer: ```suggestion
written, in Influx terms this is called
```

| Supports Sync replication                    | Yes                      | Yes        | No             |
| Supports Async replication                   | Yes                      | Yes        | Yes            |
| Concurrently evolves replicated table schema | Yes                      | Yes¹       | Yes²           |
| Same Insert / Update order on all nodes      | Yes                      | Yes        | Yes²           |
| No gaps and duplicates after failover        | Sync mode only           | Yes        | Yes            |
| Uses WAL for Replication                     | Yes                      | Yes³       | Yes            |

¹ Concurrent schema updates have to be re-tried

² InfluxDB cloud is a closed-source system, certain conclusions are made on the
assumption of the reasonable use of Kafka WAL partitioning and the correctness
of this claim depends on the implementation.

³ ClickHouse replication Data Part plays the role of WAL

To draw up some conclusions:

- All 3 systems replicate by writing to the Write Ahead Log and copying it
  across the nodes.

- Asynchronous replication, where data written to Node 1 is eventually visible

>   at Node 2, is the most popular approach used by InfluxDB Cloud and is the

Reviewer: ```suggestion
  to Node 2, is the most popular approach used by InfluxDB Cloud and is the
```

Author: I think `at Node 2` or `via Node 2` is better here. Essentially I don't care what is visible to Node 2, I care what it will reply when I query it.
